### PR TITLE
KAFKA-13702: Connect RestClient overrides response status code on request failure

### PR DESCRIPTION
**Pull request by elkkhan**
In case the submitted request status is >=400, the connect RestClient [throws](https://github.com/apache/kafka/blob/8047ba3800436d6162d0f8eb707e28857ab9eb68/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java#L133) a ConnectRestException with the proper response code, but it gets intercepted and [rethrown with 500 status code](https://github.com/apache/kafka/blob/8047ba3800436d6162d0f8eb707e28857ab9eb68/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java#L147), effectively overriding the actual failure status.

Includes a new unit-test class for the RestClient

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Why are we setting up this mock? As far as I can tell it's not used where.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: This test case is about more than just handling errors in the client, right? We might want to rename it to something like `testErrorResponseStatusGetsPreserved` and then possibly add other cases that cover `IOException`s, unexpected (i.e, 1XX or 3XX) HTTP status codes, failures during request signing, and other possible `RuntimeException`s that may be encountered in the body of `RestClient::httpRequest`.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Nit: it's a little strange that so many of these parameters are `null`. Wouldn't it be more realistic to give real values here, even if they're just constants for the sake of testing?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: It'd also be fine to leave these other cases out for now since you've strictly increased the coverage for this class with the cases you've added already, but I do think it'd still be worth adding specificity to the name of the test case here just to illustrate to future maintainers what the test covers and what it does not.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Nit: can we flip the order of arguments here and elsewhere? The `assertEquals` method should be called with the expected value as the first argument, and the actual value as the second. This comes in handy if tests fail since the error message usually says something like "Expected &lt;thing&gt; but got &lt;other thing&gt;".

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Nit: can just replace this all with `throws Exception`; makes it easier to modify these tests in the future if the set of expected exceptions changes.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Good call! Nice to have coverage for this case 👍

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: It'd give us better coverage if we performed an assertion against the actual `TestDTO` object in the response, instead of the JSON-serialized string of it.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Also, is there anything preventing us from giving a real object for the `requestBodyData` parameter? We already have JSON serialization/deserialization logic set up for the `TestDTO` object, it should be easy enough to take advantage of it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java:159]: I can see why we're catching and re-throwing this kind of exception here, and I'm also having trouble thinking of anything significantly cleaner, so this might be good enough as-is. But do you think a brief comment might help other developers understand the intent here? The control flow is a little tricky.

**Review by C0urante — COMMENTED** [review]
Thanks @elkkhan, this looks pretty good. One high-level note: we're trying to transition to using Mockito instead of EasyMock/PowerMock in our unit tests. Would it be possible to use that mocking library instead with the newly-introduced `RestClientTest` suite?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: @C0urante I'm curios about the reason for throwing a ConnectRestException with 500 error code when the client receives an unexpected 1xx or 3xx, only logging the actual 1xx/3xx code. Do you have a clue why ConnectRestException doesn't throw the actual error code? Is this an issue and should it be addressed in this PR?

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: I think a 500 in that case is intentional, since we don't know how to handle those statuses gracefully and there's nothing in the code base that would cause them to be returned. I haven't seen any issues from this logic, either personally or on the Jira for the project. It's probably fine to leave as-is.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: true, thanks for noticing - I don't remember anymore what was the intent behind it :) 
removed

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java
Comment (elkkhan) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java:159]: sure, added a comment, let me know if the wording is understandable

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: I've added tests for the cases you've mentioned.
sending the request may trigger 3 different exceptions, so i've put them in a parameterised test, but in junit 4 you can't have a single parameterised test case - the whole class runs once for each parameter.
I've used [this](https://stackoverflow.com/a/35057629) approach to isolate that test case, but IMO the test class became a little hard to read.

Do you know any better way to isolate that one parameterised test case? I know JUnit5 can do that but it's not in the classpath unfortunately

there could also just be 3 separate test cases, but a parameterised test seems like the perfect fit here

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: replaced nulls with real values

**Comment by elkkhan** [comment]
@C0urante thanks for the review, I've addressed your comments

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java:159]: 👍 looks good to me, thanks!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: We should probably change this to either `HmacSHA1` or `HmacSHA256`, since those are the only two algorithms guaranteed to be present on JVMs that comply with the Java platform specs. See the [Mac Javadocs](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/javax/crypto/Mac.html) for reference.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Can we use the `@Mock` annotation here? And if not, is it safe to establish a `static` field here, which will be reused across test cases?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Yeah, parameterized tests would be a great fit for this case. One workaround I've adopted in a recent PR is extracting the common logic of the test case into a reusable method, then writing individual test cases that each delegate to that reusable method with different arguments. For an example, see the `DistributedHerderTest` suite: https://github.com/apache/kafka/blob/277c4c2e97d2aef096ab0e998fdd1ae513508798/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java#L2976-L3050

It shouldn't be too much work to convert these tests to that style, but it's up to you if you think it'd be more or less readable. I think the style you have right now is acceptable as-is, though obviously it'd be better if we could replace it with JUnit 5-style generated test cases.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: Is it necessary to construct a real key here, or can we use a mocked key? I'm worried this may not be compatible with a wide range of operating systems/JVMs.

**Review by C0urante — COMMENTED** [review]
Looking pretty good! Left a few more comments but assuming they're addressed, this looks good to me. It does look like the CI builds are failing with Checkstyle errors, though; can you take a look at those?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: @C0urante would it be too much overhead to add junit-jupiter-params to classpath? if not, I'd be happy to refactor this to use JUnit 5 parameterized tests, otherwise I think it can be left as is

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: `@Mock` can't be used since it needs a `MockitoJUnitRunner` - we already have the `Parameterized` runner here and we can't have multiple runners on a class. Good point about the `static` mock, I missed that.

`Any static field saves its state for the duration of the JVM's execution (unless code changes its value, of course). JUnit uses one JVM for all of its tests, so, yes, static fields save state between tests.`

Changing the field to non-static

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: replaced with a mock key

**Comment by elkkhan** [comment]
@C0urante addressed the comments, I think the only outstanding one is whether we should add junit-jupiter-params to dep. list and use that for parameterised tests

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java:127]: We should also make sure that the cause of the exception is properly tracked:
```suggestion
            assertIsInternalServerError(e);
            assertEquals(requestException, e.getCause());
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java:118]: Because we don't set up any expectations for `Request::header`, we get a `NullPointerException` right now at this line during these tests, instead of the exception we want to be thrown from `Request::send`: https://github.com/apache/kafka/blob/6495a0768cae1086d4b1dfa466967dfe178c1553/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/InternalRequestSignature.java#L57-L58

Luckily, there's an easy fix for this:
```suggestion
            when(req.send()).thenThrow(t);
            when(req.header(anyString(), anyString())).thenReturn(req);
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: We can still use the `@Mock` annotation by tweaking this part:

```suggestion
        @Rule
        public MockitoRule initRule = MockitoJUnit.rule();

        @Mock
        private HttpClient httpClient;
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: I left a comment on the latest commit about how we can still use the `@Mock` annotation; hope this helps.

**Review by C0urante — COMMENTED** [review]
Almost there! Noticed a small issue with the testing logic while experimenting locally with the `@Mock` annotation question.

There's also still Checkstyle issues that are causing the build to fail; you can see these at the bottom of the page by following the links under the `Some checks were not successful` header.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: I think we should hold off on using Junit 5 for now; these changes are fine as-are.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java:109]: @C0urante I believe this is already in the latest commit, any chance you were referring to a the previous commit?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java]: nice, thanks - didn't know this. fixed

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java:127]: fixed

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java
Comment (elkkhan) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestClientTest.java:118]: nice catch, thx.  setupHttpClient does this but I forgot to do it here and didn't notice the NPEs.
fixed

**Comment by elkkhan** [comment]
@C0urante thanks, addressed the comments. hopefully checkstyle is all good now - I wasn't able to run the gradle task locally for some reason, it kept getting stuck, but now it's working and reporting no issues

**Review by C0urante — APPROVED** [review]
LGTM, thanks @elkkhan.

@mimaison if you have time can you take a look at this quick fix? Would be nice to include in the 3.3.0 release if possible.

**Review by mimaison — APPROVED** [review]
LGTM, thanks for the PR

**Pull request merged**
